File: src/sample02/Conta.java
package sample02;

public class Conta {
    String numero;
    double saldo;

    Conta(String num, double saldoInicial) {
        this.numero = num;
        this.saldo = saldoInicial;
    }

    boolean sacar(double valor) {
        if ((saldo - valor) < 0) {
            return false;
        }
        saldo -= valor;
        return true;
    }

    void depositar(double valor) {
        saldo = saldo + valor;
    }

    double consultarSaldo() {
        return saldo;
    }

    boolean transferir(Conta destino, double valor) {
        if (!this.sacar(valor)) {
            return false;
        }

        destino.depositar(valor);
        return true;
    }

    @Override
    public String toString() {
        return "Conta{" +
                "numero='" + numero + '\'' +
                ", saldo=" + saldo +
                '}';
    }
}

File: src/sample09/conta/Conta.ts
import { ValorInvalidoError } from "./Errors";

export class Conta {
  private _numero: String;
  private _saldo: number = 0;

  constructor(numero: String, saldoInicial: number) {
    this._numero = numero;
    this.depositar(saldoInicial);
  }

  get numero(): String {
    return this._numero;
  }

  get saldo(): number {
    return this._saldo;
  }

  sacar(valor: number): void {
    this.validarValor(valor);

    if (this._saldo >= valor) {
      this._saldo = this._saldo - valor;
    }
  }

  depositar(valor: number): void {
    this.validarValor(valor);
    this._saldo = this._saldo + valor;
  }

  transferir(contaDestino: Conta, valor: number): void {
    this.sacar(valor);
    contaDestino.depositar(valor);
  }

  private validarValor = (valor: number) => {
    if (valor <= 0) {
      throw new ValorInvalidoError();
    }
  }
}
